lib/util/Components.js

```diff
@@ -27,7 +27,12 @@ function usedPropTypesAreEquivalent(propA, propB) {
     if (!propA.allNames && !propB.allNames) {
       return true;
     }
-    if (Array.isArray(propA.allNames) && Array.isArray(propB.allNames) && propA.allNames.join('') === propB.allNames.join('')) {
+    if (Array.isArray(propA.allNames)
+      && Array.isArray(propB.allNames)
+      && (
+        (propA.allNames === propB.allNames)
```

Comment:
when will this check ever actually be true?

Reply:
From my testing, most of the time.
This ended up being a strict optimisation above and beyond the (expensive) joins.
Ideally, we'd probably be better off swapping out the join comparison for some kind of fast array of strings equality check as it's still run a good chunk of the time.

Please check my logic below:

With this basic benchmarking on a project with about 550 jsx files:
```
let conditionChecked = 0
let arraysDoHaveIdentityEquality = 0
let arraysDoNotHaveIdentityEquality = 0
function usedPropTypesAreEquivalent(propA, propB) {
  if (propA.name === propB.name) {
    if (!propA.allNames && !propB.allNames) {
      return true;
    }
    if (Array.isArray(propA.allNames) && Array.isArray(propB.allNames)) {
      conditionChecked ++
    }
    if (Array.isArray(propA.allNames) && Array.isArray(propB.allNames) && propA.allNames.join('') === propB.allNames.join('')) {
      if(propA.allNames === propB.allNames) {
        arraysDoHaveIdentityEquality ++
      } else {
        arraysDoNotHaveIdentityEquality ++
      }
      console.log('Arrays have identity equality:', arraysDoHaveIdentityEquality)
      console.log('Arrays do not have identity equality', arraysDoNotHaveIdentityEquality)
      console.log('Ratio', arraysDoHaveIdentityEquality / arraysDoNotHaveIdentityEquality)
      console.log('Total times condition reached', arraysDoHaveIdentityEquality + arraysDoNotHaveIdentityEquality)
      console.log("Condition checked", conditionChecked)
      return true;
    }
    return false;
  }
  return false;
}
```

With an otherwise unaltered lib/util/Components.js,
I sometimes got a ratio of arrays having identity equality of 4 to 1 when the condition had been checked nearly 80,000 times:
```
Arrays have identity equality: 60555
Arrays do not have identity equality 14723
Ratio 4.112952523262922
Total times condition reached 75278
Condition checked 78661
```

but this settled down to about 2 to 1 after the condition had been checked almost a million times.
```
Arrays have identity equality: 572805
Arrays do not have identity equality 291390
Ratio 1.9657675280551838
Total times condition reached 864195
Condition checked 990195
```

Reply:
I guess I'm surprised that the AST parser would generate two memoized `allNames` arrays for different nodes.